File: pitcher-service/src/integrationTest/java/io/pitchcast/pitcherservice/web/pitchercontroller/PitcherController_addIT.java
package io.pitchcast.pitcherservice.web.pitchercontroller;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.github.benas.randombeans.api.EnhancedRandom;
import io.pitchcast.pitcherservice.domain.Pitcher;
import io.pitchcast.pitcherservice.service.PitcherService;
import io.pitchcast.pitcherservice.web.PitcherController;
import io.pitchcast.pitcherservice.web.dto.PitcherDto;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.http.MediaType;
import org.springframework.test.context.junit.jupiter.SpringExtension;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;
import org.springframework.test.web.servlet.result.MockMvcResultMatchers;

import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
import static org.mockito.Matchers.any;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

@ExtendWith(SpringExtension.class)
@WebMvcTest(controllers = PitcherController.class)
public class PitcherController_addIT {

    @Autowired
    private MockMvc mockMvc;

    @Autowired
    private ObjectMapper objectMapper;

    @MockBean
    private PitcherService pitcherService;

    @Test
    public void shouldReturn201IfPitcherWasAdded() throws Exception {
        // given
        ArgumentCaptor<Pitcher> pitcherArgumentCaptor = ArgumentCaptor.forClass(Pitcher.class);

        PitcherDto validPitcher = EnhancedRandom.random(PitcherDto.class);

        // then
        mockMvc.perform(
                MockMvcRequestBuilders.post("/pitcher/")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(validPitcher))
        ).andExpect(MockMvcResultMatchers.status().is(201));

        verify(this.pitcherService).savePitcher(pitcherArgumentCaptor.capture());

        Pitcher sentPitcher = pitcherArgumentCaptor.getValue();

        assertThat(sentPitcher).isEqualToIgnoringGivenFields(validPitcher, "pitcherId", "handed");
        assertThat(sentPitcher.getHanded().name()).isEqualTo(validPitcher.getHanded().name());

    }

    @Test
    public void shouldReturn400IfPitcherHasNullValues() throws Exception {

        PitcherDto validPitcher = new PitcherDto();

        // then
        mockMvc.perform(
                MockMvcRequestBuilders.post("/pitcher/")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(validPitcher))
        ).andExpect(MockMvcResultMatchers.status().is(400));

        verify(pitcherService, times(0)).savePitcher(any());
    }

}

File: pitcher-service/src/integrationTest/kotlin/io/pitchcast/pitcherservice/web/PitcherController_addIT.kt
package io.pitchcast.pitcherservice.web

import com.fasterxml.jackson.databind.ObjectMapper
import io.github.benas.randombeans.api.EnhancedRandom
import io.pitchcast.pitcherservice.domain.Pitcher
import io.pitchcast.pitcherservice.service.PitcherService
import io.pitchcast.pitcherservice.web.dto.PitcherDto
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.mockito.ArgumentCaptor
import org.mockito.ArgumentMatchers
import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultMatchers
import testing.WebMvcSliceTest

@WebMvcSliceTest(forController = [PitcherController::class])
class PitcherController_addIT {

    @Autowired lateinit var mockMvc: MockMvc
    @Autowired lateinit var objectMapper: ObjectMapper
    @MockBean lateinit var pitcherService: PitcherService

    @Test fun `should return 201 if pitcher was added successfully`() {
        // given
        val pitcherArgumentCaptor = ArgumentCaptor.forClass(Pitcher::class.java)

        val validPitcher = EnhancedRandom.random(PitcherDto::class.java)

        // then
        mockMvc.perform(
                MockMvcRequestBuilders.post("/pitcher/")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(validPitcher))
        ).andExpect(MockMvcResultMatchers.status().`is`(201))

        verify<PitcherService>(this.pitcherService).savePitcher(pitcherArgumentCaptor.capture())

        val sentPitcher = pitcherArgumentCaptor.value

        assertThat(sentPitcher).isEqualToIgnoringGivenFields(validPitcher, "pitcherId", "handed")
        assertThat(sentPitcher.handed.name).isEqualTo(validPitcher.handed.name)

    }

    @Test fun `should return 400 if pitcher has null values`() {

        val validPitcher = PitcherDto()

        // then
        mockMvc.perform(
                MockMvcRequestBuilders.post("/pitcher/")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(validPitcher))
        ).andExpect(MockMvcResultMatchers.status().`is`(400))

        verify<PitcherService>(pitcherService, times(0)).savePitcher(ArgumentMatchers.any())
    }

}